### ISSUE TITLE
Scheduler V2

### ISSUE THREAD

**Issue by lmars** [flynn/flynn#1484]
The current scheduler implementation feels slightly hacked together and there are a number of open issues relating to it's short comings (#1182, #1434, #686, #129, #54, #1340).

We should refactor the scheduler to both fix these issues and make it more flexible to handle different scheduling policies (e.g. resource based, tag limiting etc.).

Here are the main responsibilities of the scheduler:

* Keep an accurate state of running jobs in memory (being handled in #1373), and ensure it matches the requested scale
* React to formation changes by adding / removing jobs from hosts
* Restart crashed jobs using an exponential backoff policy
* Update the `jobs_cache` DB table with job states (including ones scheduled to start in the future)

Some improvements to be made based on the current implementation:

* Support arbitrary constraints when running new jobs (e.g. available resources, host tags, data locality etc.)
* Actively keep running jobs in sync with formations, using a deterministic algorithm to determine which jobs to remove if necessary (so if multiple schedulers are scheduling, they will both start new jobs then race to stop the same duplicate job, thus minimising the impact of #1182)
* Actively keep `jobs_cache` table in sync with cluster state
* When the leader crashes, maintain the backoff policy from the crashed scheduler
* Simplify "omni" jobs by modifying formation process counts on host change so that they always match the actual number of jobs of that type (it is confusing that `scheduler=2` could mean "2 jobs per host" rather than just "2 jobs in the cluster")
* Make unit testing simpler (all scheduler testing is currently at the integration level)

/cc @titanous

**Comment by heavenlyhash** [comment]
Re: "if multiple schedulers are scheduling", maybe it would be easier to maintain consistency if schedulers only take actions driven by consulting (and immediately atomically updating) the actual consensus system?  If schedulers become stateless outside of the consensus state, then it doesn't matter how many of them there are participating.  After all, as was wisely said in #1373, making it "appear that there is a consistent view of the cluster, when that is not actually true" is a source of complications.  It might be possible to engineer schedulers to have local state with inconsistent views of the world and anneal toward consistency anyway, but it sounds a lot harder.  Rather than guiding schedulers towards races in their work, the race can be shuffled to transactions in the consensus system, and as a bonus, become more monitorable.

**Comment by lmars** [comment]
@heavenlyhash I agree (I actually proposed [here](https://github.com/flynn/flynn/issues/1373#issuecomment-87486340) using service meta to communicate on the cluster state), but I guess my concern is the scheduler depending on services that it itself schedules to be able to do it's job (e.g. if discoverd goes down, the scheduler can't restart it as it can't come to consensus without it, ditto for postgres).

That being said I could be convinced that discoverd going down is an edge case, and that the benefits of using say CAS service metadata outweigh the ad-hoc code to handle a discoverd outage.

**Comment by titanous** [comment]
I'm not sure routing the state through a consistent datastore improves anything. The source of truth is the individual hosts in the cluster, and schedulers should try to stay as close to that as possible by using discovery and event streams.

Having multiple schedulers that think they are the leader is an edge case that we should try to avoid. We don't need to optimize for it, we just need to make sure that if it happens the cluster doesn't break and it is resolved as quickly as possible.

**Comment by kwando** [comment]
I'd like the automatic restart backoff timeout to be adjustable per job.

**Comment by lmars** [comment]
We need to make sure the new scheduler correctly handles #1366 (i.e. when a job is scheduled to start in the future, scale events should reset the timers and possibly send a down job event).

**Comment by jzila** [comment]
Currently if multiple schedulers believe they are leaders, the world somehow ends up having (num leaders * formation) number of processes. If the new scheduler is scheduling correctly, even having multiple leaders would trend towards the correct world state:
- scale web 1=>3
- schedulers 1 and 2 both attempt to scale to 3 (delta == 2)
- each scheduler will be notified of up events, and realize that the state of the world is incorrect (6 instances, desired=3).
- at this point, the schedulers will most likely thrash between 0 and 6 instances, maybe eventually landing on the correct scale.

I have two proposed solutions:

1.  A scheduler notices thrashing and reduces its aggressiveness: instead of attempting to reconcile to desired state in 1 step, it backs off. So, if it notices that delta between desired and actual state is 3, and its last action had a delta of 2, its next action will only have a delta of 1. It will then take its next action after some timeout. So with the above scenario, the formation might thrash like so: 1->6->4(->2->4)*->3. I'm assuming that the race between the multiple schedulers will eventually result in one of them winning and the others not taking an action.
1.  If a system does not reach desired state within the scale timeout, a leader election is triggered.
  1. A new leader will always attempt to re-establish the Correct world state based on its understanding of desired state.

I like the second one better since it's much simpler. Not sure if there's large overhead to a leader election.

**Comment by titanous** [comment]
> at this point, the schedulers will most likely thrash between 0 and 6 instances, maybe eventually landing on the correct scale.

I think this can be resolved by ensuring that scheduler-triggered job terminations are deterministically chosen (based on objective criteria). That way both of the schedulers would end up attempting to terminate the same jobs.

**Comment by jzila** [comment]
What do you guys think of factoring out the controller client into an interface, and then implementing a fake controller client that we can use in unit tests? This would simplify a lot of testing, enable us to instantiate one without all the `http.Client` overhead, and allow us to build custom stubs for whatever we want to test.

**Comment by titanous** [comment]
You should be able to make an interface that has the subset of methods
you need without turning the entire client into an interface. Large
interfaces are an anti pattern and I'm trying to get away from them.

**Comment by jzila** [comment]
OK will do.

**Comment by jzila** [comment]
I've been going through the existing scheduler code, and it looks like the current method for removing jobs is almost correct, but not quite. Currently it seems to be written to remove the oldest jobs first ([but actually removes the newest jobs first](https://github.com/flynn/flynn/blob/master/controller/scheduler/main.go#L885)). I'm also guessing that the scheduler doesn't currently sync as often as it should, hence why it's not repairing the formations correctly.

**Comment by lmars** [comment]
Some outstanding changes:

- [ ] fix [this TODO](https://github.com/flynn/flynn/blob/scheduler-v2-base/controller/scheduler-v2/scheduler.go#L218) (i.e. just set `diff` to `ef.Processes` if the formation is new)
- [ ] update `Scheduler.Sync` (or add `Scheduler.SyncFormations`) to check in-memory formations are accurate (i.e. we haven't missed a scaling event)
- [ ] update `Scheduler.Sync` to check that the jobs running in the cluster are what is expected based on known formations
- [ ] stream job events from all hosts, modify in-memory state, call `PutJob`
- [ ] restart crashed jobs using an exponential backoff policy, make this visible in the controller (e.g. a `scheduled` state and a `run_at` column) and update `Scheduler.Sync` to ensure the run_at time is accurate (so if a scheduler gets promoted, it knows that certain jobs need to be started in the future)
- [ ] handle leader demotion (basically just stop starting / stopping jobs, but continue to track job / formation state and calling `PutJob`)
- [ ] watch for host changes and update "omni" process types by modifying formation counts to match the number of active hosts

We should make these changes as incremental pull requests into the [scheduler-v2-base](https://github.com/flynn/flynn/tree/scheduler-v2-base) branch.

/cc @jzila

**Comment by jzila** [comment]
Will do. Thanks!

**Comment by jzila** [comment]
@lmars ready for review.

**Comment by lmars** [comment]
If for any reason there are no hosts available (e.g. an update is running and hosts drop temporarily from service discovery), we should try multiple times to pick a host to run a job on rather than giving up (the current scheduler actually panics :scream:)

/cc @jzila

**Comment by jzila** [comment]
@lmars Any thoughts as to how to communicate changes to controller jobs back to the scheduler? Should I create a `StreamJobs` allowing any `PutJob` calls to cascade back to each scheduler? How should we handle the fact that all the schedulers are calling `PutJob`, and thus each scheduler is calculating its own `RunAt` number for exponential decay? Should only the leader handle crashed processes?

**Comment by lmars** [comment]
@jzila 

> Any thoughts as to how to communicate changes to controller jobs back to the scheduler? Should I create a StreamJobs allowing any PutJob calls to cascade back to each scheduler?

The job states in the controller are just a cache of the actual cluster state, and since the schedulers will be keeping in sync with the cluster, there is no need to stream the cache back into the scheduler.

> How should we handle the fact that all the schedulers are calling PutJob, and thus each scheduler is calculating its own RunAt number for exponential backoff? Should only the leader handle crashed jobs?

Yes only the leader should be scheduling the restart of crashed jobs (and cacheing the `RunAt` value that it calculates in the controller).

**Comment by jzila** [comment]
@lmars but `SyncJobs` syncs with the cluster, not the controller. If we want to sync against the controller, then shouldn't we also be streaming changes to the controller jobs? I.e. we don't want SyncJobs to do anything that isn't already done by streaming tasks, because its purpose is anti-entropy.

**Comment by lmars** [comment]
@jzila I'm not sure I understand what you are asking.

The schedulers each stream changes from hosts in an attempt to keep an accurate state of the cluster in-memory, and they periodically call `SyncJobs` in case they miss events.

The schedulers also call `PutJob` anytime they see a job state change (either from a host stream or a call to `SyncJobs`) so the controller has a cache of jobs and their current state.

Given this, why would the schedulers need to stream that cache back into their in-memory state? The way I see it, information only needs to flow one-way from hosts -> scheduler -> controller.

**Comment by jzila** [comment]
@lmars I'm asking how the non-leader schedulers will pull updates to RunAt, so that if the leader gets demoted, the new leader knows when to resume the job execution attempts.

**Comment by titanous** [comment]
I'm not convinced that RunAt is so critical that we need to work hard to replicate it. What happens if we just don't?

**Comment by jzila** [comment]
The leadership change will assert each formation, so if there are crashed jobs on the formation it'll try to start jobs immediately upon leadership change. Then the new scheduler will run through its exponential backoff.

TL;DR: No RunAt -> a leadership change will reset any exponential backoff.

**Comment by titanous** [comment]
@jzila I don't think resetting the backoff is a problem.

**Comment by jzila** [comment]
@titanous Cool, let's go without `RunAt` then, and handle exponential backoff only on the leader.

**Comment by lmars** [comment]
@jzila ok I see your point.

So thinking about this some more and looking at the current scheduler, when it becomes leader it pulls all the jobs from the controller and marks any as down which aren't actually running (see [here](https://github.com/flynn/flynn/blob/master/controller/scheduler/main.go#L240-L275)). We should also be doing that in the V2 scheduler to make sure stopped jobs don't appear in `flynn ps`, and maybe we should be doing that as an anti-entropy thing (rather than just when there is a new leader).

If we do do that, then syncing `RunAt` can be done at the same time.

WDYT?

**Comment by lmars** [comment]
If `host.AddJob` fails (e.g. because the `flynn-host` daemon has crashed but has yet to be removed from service discovery), we should retry scheduling the job (which should hopefully pick a healthy host).

/cc @jzila

**Comment by lmars** [comment]
When a host goes offline (i.e. it is removed from service discovery), we should explicitly stop streaming job events from that host (the streams currently auto-reconnect).

/cc @jzila

**Comment by jzila** [comment]
@lmars I agree with @titanous that job restart state can be maintained in the leader, and there's no real benefit to syncing that across schedulers. Regarding `host.AddJob` failing, I intend to do that as part of the same exponential backoff system that we'll use for starting jobs.

Finally, regarding stopping job streams, how will we know that a host is now offline?

**Comment by lmars** [comment]
@jzila 

> how will we know that a host is now offline?

We will get a down event from discoverd.

**Comment by jzila** [comment]
@titanous meaning from `StreamHosts`?

**Comment by lmars** [comment]
@jzila yes, currently StreamHosts [drops the down events](https://github.com/flynn/flynn/blob/master/pkg/cluster/client.go#L92-L95) so we will need to modify that so they are passed through somehow.

**Comment by jzila** [comment]
@lmars there's also the issue that StreamHosts doesn't communicate host events; it communicates host clients. So we're actually talking about changing the interface.

**Comment by jzila** [comment]
Updated To Do list:
- [ ] Implement exponential backoff for starting/re-starting jobs. Store the restart state only on the leader. Failover to a new leader will restart the exponential backoff counter for any jobs that are still trying to be started. Cap exponential backoff according to #1340.
- [ ] Refactor `StreamHosts` to `StreamHostEvents` (https://github.com/flynn/flynn/commit/519716313df4806003ce81775510913d9e344d54#diff-1708b7928ae9bd10d26b695ccece8bf4)
- [ ] Refactor `hostClient.StreamEvents` to each stream to its own channel, and multiplex the streams into 1 channel.
- [ ] Keep `s.hosts` in sync with `s.Hosts()`: handle closing host job streams when we lose a host.
- [ ] Add tests for more complex scenarios:
 - [ ] Multiple schedulers (1 leader); verify that each maintains correct state, and verify that leadership transfer works correctly.
 - [ ] Exponential backoff
 - [ ] Error first time starting a job (Attempt to start a job with no hosts)
 - [ ] Error first time stopping a job (Attempt to stop a nonexistent job)
 - [ ] Start Omni job with at least 2 hosts

**Comment by jzila** [comment]
FYI current status: many races happening, mostly regarding intermittent job state: the time between when the scheduler kicks off a host action and when the results of that action are received is currently resulting in dupe actions being launched. I have to figure out a clean way to mitigate this issue.

**Comment by lmars** [comment]
@jzila still having issues with races? I was hoping by having a single threaded loop that things would be predictable, let me know if you want me to take a look at anything, or grab me in IRC.

**Comment by lmars** [comment]
We a new host comes online, we should rebalance jobs of each type (e.g. if we cycle a host in a 3 node cluster, and a service was scaled to 3, we will end up with one host running two of those jobs, even after the node comes back online)

**Comment by jzila** [comment]
@lmars let's call that an enhancement, I'm not sure whether that feature is necessary for 2.0.
